### PR TITLE
Feat: APP-2376 - Implement primary delegation flow

### PR DESCRIPTION
**Pull request by cgero-eth**
## Description

- Use `VITE_FEATURE_FLAG_DELEGATION` flag to enable / disable the delegation flow
- Implement primary delegation flow from the governance page

Task: [APP-2376](https://aragonassociation.atlassian.net/browse/APP-2376)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran all tests with success and extended them if necessary.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.
- [x] I have tested my code on the test network.

[APP-2376]: https://aragonassociation.atlassian.net/browse/APP-2376?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/containers/delegateVotingMenu/delegateVotingForm.tsx [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  const isLoading = status === 'loading';
+  const isError = status === 'error';
+
+  const delegateLabel = getDelegateLabel(isReclaimMode, isLoading, isError);
+  const ctaIcon = isLoading ? (
+    <Spinner size="xs" color="white" />
+  ) : isError ? (
+    <IconReload />
+  ) : undefined;
+  const alertLabel = isReclaimMode
```
Comment (Fabricevladimir) [inline src/containers/delegateVotingMenu/delegateVotingForm.tsx:115]: Give me a space! lol

File: src/containers/delegateVotingMenu/delegateVotingForm.tsx [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+          />
+        )}
+      </FormGroup>
+    </div>
+  );
+};
+
+const FormGroup = styled.div.attrs((props: {classname?: string}) => ({
+  className: `flex flex-col gap-1.5 ${props.classname}`,
+}))<{classname?: string}>``;
```
Comment (Fabricevladimir) [inline src/containers/delegateVotingMenu/delegateVotingForm.tsx]: If you cast className as a string, then you can make use of the usual `className` react attribute on the styled component without any issues.

```suggestion
const FormGroup = styled.div.attrs({
  className: 'flex flex-col gap-1.5' as string,
})``;
```

File: src/containers/delegateVotingMenu/delegateVotingMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    if (isConnected && isDelegateVotingOpen && tokenBalance?.value === 0n) {
+      open('gating');
+      close('delegateVoting');
+    }
+  }, [isConnected, tokenBalance?.value, isDelegateVotingOpen, close, open]);
+
+  if (!isConnected && isDelegateVotingOpen) {
+    return <LoginRequired isOpen={true} onClose={handleCloseLogin} />;
+  }
+
```
Comment (Fabricevladimir) [inline src/containers/delegateVotingMenu/delegateVotingMenu.tsx:165]: Sepehr is working on a PR that should make this unnecessary all over the place I think.

File: src/context/globalModals.tsx [hunk 4]
```diff
@@ -79,100 +82,64 @@ const GlobalModalsProvider: React.FC<Props> = ({children}) => {
     useState<GlobalModalsContextType['isPoapClaimOpen']>(false);
   const [isExportCsvOpen, setIsExportCsvOpen] =
     useState<GlobalModalsContextType['isExportCsvOpen']>(false);
+  const [isDelegateVotingOpen, setIsDelegateVotingOpen] =
```
Comment (Fabricevladimir) [inline src/context/globalModals.tsx:85]: I was really hoping this context would die off or at the very least we get an option to pass a callback onClose. Sometimes you really need that flexibility, especially with the login flow. When you try to do an action, the login menu will pop up, once you've logged in, you have to retry the action. Leads to an unnecessary click. Anyway food for thought.

File: src/hooks/useDaoToken.tsx [hunk 5]
```diff
@@ -16,11 +16,11 @@ export function useDaoToken(
   const pluginClient = usePluginClient('token-voting.plugin.dao.eth');
 
   useEffect(() => {
-    async function getDaoMetadata() {
+    async function getDaoMetadata(address: string) {
```
Comment (Fabricevladimir) [inline src/hooks/useDaoToken.tsx]: Not sure when this copy pasta happened, but can you update the name since this isn't getting DAO metadata?

File: src/locales/en/common.json [hunk 6]
```diff
@@ -1358,4 +1358,4 @@
     "labelYourDelegate": "Your delegate",
     "labelYou": "You"
   }
-}
\ No newline at end of file
+}
```
Comment (Fabricevladimir) [inline src/locales/en/common.json:1361]: -_-

File: src/containers/delegateVotingMenu/delegateVotingSuccess.tsx [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            label={t('modal.delegation.successCtaLabel')}
+            mode="primary"
+            onClick={handleCommunityClick}
+          />
+          <ButtonText
+            label={t('modal.delegation.successBtnSecondaryLabel')}
+            mode="secondary"
+            iconRight={<IconLinkExternal />}
+            onClick={handleTransactionClick}
+          />
```
Comment (Fabricevladimir) [inline src/containers/delegateVotingMenu/delegateVotingSuccess.tsx]: Both of these buttons should be of size large

File: src/containers/delegateVotingMenu/delegateVotingForm.tsx [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+          isActive={isLoading}
+          mode="primary"
+          iconLeft={ctaIcon}
+          onClick={onDelegateTokens}
+        />
+        <ButtonText
+          className="w-full"
+          label={t('labels.cancel')}
+          mode="secondary"
+          onClick={handleCancel}
```
Comment (Fabricevladimir) [inline src/containers/delegateVotingMenu/delegateVotingForm.tsx]: Can you add `size="large"` to these buttons as well?

**Review by Fabricevladimir — DISMISSED** [review]
Thanks @RuggeroCino, lots of good things. Left a few comments. Additionally, noticed a weird issue when we try to reclaim voting power after having delegated it to another wallet:

https://github.com/aragon/app/assets/16927999/c3f39cc5-4d91-48c3-bd30-16425e97d0cf

File: src/containers/delegateVotingMenu/delegateVotingForm.tsx (hunk 1 above)
Comment (cgero-eth) [reply src/containers/delegateVotingMenu/delegateVotingForm.tsx:115]: Ahaah will add some spaces 🙈

File: src/containers/delegateVotingMenu/delegateVotingForm.tsx [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    onDelegateChange({
+      address: address as string,
+      ensName: ensName != null ? ensName : '',
+    });
+  };
+
+  const isDelegateValid =
+    delegate.address != null &&
+    delegate.address !== '' &&
+    delegate.address !== currentDelegate;
```
Comment (Fabricevladimir) [inline src/containers/delegateVotingMenu/delegateVotingForm.tsx]: Not sure if the issue is here, but when the connected wallet ens is what shows up as the delegate, the form allows me to run a transaction to reclaim my delegation even though I hold the voting power. 

Also happens if I paste my wallet address into the form while "Delegate to somebody" is selected.

File: src/containers/delegateVotingMenu/delegateVotingForm.tsx (hunk 2 above)
Comment (cgero-eth) [reply src/containers/delegateVotingMenu/delegateVotingForm.tsx]: Oh nice suggestion!

File: src/locales/en/common.json (hunk 6 above)
Comment (cgero-eth) [reply src/locales/en/common.json:1361]: Auto formatting 😅

File: src/context/globalModals.tsx (hunk 4 above)
Comment (cgero-eth) [reply src/context/globalModals.tsx:85]: We can for sure improve this dialog handling and pass custom properties to dialogs, it makes sense to open a tech debt for that 👍🏻

File: src/containers/delegateVotingMenu/delegateVotingMenu.tsx (hunk 3 above)
Comment (cgero-eth) [reply src/containers/delegateVotingMenu/delegateVotingMenu.tsx:165]: Mmm.. I'm not sure if we can reuse the logic of the protected routes here, or at least we'll need to update Sepehr's PR to be able to customise the `handleCloseLogin` callback because on the protected routes we currently do a `navigate(-1)` while here we have a different logic :/
I'd suggest to tackle this together with the ticket to refactor the dialog handling

File: src/containers/delegateVotingMenu/delegateVotingForm.tsx (hunk 9 above)
Comment (cgero-eth) [reply src/containers/delegateVotingMenu/delegateVotingForm.tsx]: I suppose that it wasn't working as expected because I wasn't checking for "normalised" addresses, I updated the check

**Comment by cgero-eth** [comment]
> Thanks @RuggeroCino, lots of good things. Left a few comments. Additionally, noticed a weird issue when we try to reclaim voting power after having delegated it to another wallet:
> 
>  Aragon.App.-.Google.Chrome.2023-08-18.09-13-05.mp4

I updated the component to use `react-hook-form` and correctly validate the delegate address, after that I could not replicate the issue you mentioned. Let me know if this is working properly for you now

File: src/services/aragon-sdk/aragon-sdk-service.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    client?: TokenVotingClient
+  ): Promise<AsyncGenerator<DelegateTokensStepValue>> => {
+    invariant(client != null, 'delegateTokens: client is not defined');
+    const data = client.methods.delegateTokens(params);
+
+    return data;
+  };
+}
+
+export const aragonSdkService = new AragonSdkService();
```
Comment (jamesej) [inline src/services/aragon-sdk/aragon-sdk-service.ts]: Ooh I don't love this. Seems like you're adding a wrapper around what's already a wrapper just to test for whether the client is null. Can't you get rid of this class? Seems like an unnecessary indirection.

**Review by jamesej — CHANGES_REQUESTED** [review]
Great work apart from the one comment I had.

File: src/services/aragon-sdk/aragon-sdk-service.ts (hunk 10 above)
Comment (cgero-eth) [reply src/services/aragon-sdk/aragon-sdk-service.ts]: Hey @jamesej, thanks for checking the code! Sure, I usually use these `-service` classes to collect the functions and utilities (when needed) to interact with a third party service. Given that currently this class doesn't have utilities or some specific logic that can be shared across the service, I can just remove them and move these functions on the `useDelegateTokens` and `useDelegatee` hooks

**Review by jamesej — APPROVED** [review]
Thanks Ruggero!

**Review by jamesej — APPROVED** [review]
Thanks Ruggero!
